lib/IRGen/IRGen.cpp

```diff
@@ -499,7 +499,11 @@ swift::createTargetMachine(IRGenOptions &Opts, ASTContext &Ctx) {
 
   // Create a target machine.
   auto cmodel = CodeModel::Default;
-  if (Triple.isWindowsCygwinEnvironment())
+
+  // On Windows 64 bit, dlls are loaded above the max address for 32 bits.
+  // This means that a default CodeModel causes generate code to segfault
+  // when run.
+  if (Triple.isArch64Bit() && Triple.isOSWindows())
```

Comment:
It's unclear, I'm taking this from a comment in that Pr:
> Cygwin should use large memory model as default.(This may be changed
if someone ports to 32bit)

I think for now specialising this to 64 bit, which is the only one that's been tested, is the right way to go. 

Logically speaking, I'm not sure how referencing a 32 bit address on a 32 bit architecture would be bad, but I can't test this